File: src/main/java/at/irian/webstack/support/data/PaginationController.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package at.irian.webstack.support.data;

import com.avaje.ebean.Page;
import com.avaje.ebean.PagingList;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Future;

/**
 * @author Werner Punz (latest modification by $Author$)
 * @version $Revision$ $Date$
 */

public class PaginationController<T> {
    PagingList<T> _delegate;
    int lastPageAccessed = 0;

    public PaginationController(PagingList<T> delegate) {

        _delegate = delegate;
    }

    public void refresh() {
        _delegate.refresh();
    }

    public PagingList<T> setFetchAhead(boolean b) {
        return _delegate.setFetchAhead(b);
    }

    public Future<Integer> getFutureRowCount() {
        return _delegate.getFutureRowCount();
    }

    public List<T> getAsList() {
        return _delegate.getAsList();
    }

    public int getPageSize() {
        return _delegate.getPageSize();
    }

    public int getTotalRowCount() {
        return _delegate.getTotalRowCount();
    }

    public int getTotalPageCount() {
        return _delegate.getTotalPageCount();
    }

    public Page<T> getPage(int i) {
        lastPageAccessed = i;
        return _delegate.getPage(i);
    }

    public Page<T> getPage() {
        return _delegate.getPage(lastPageAccessed);
    }

    public int getLastPageAccessed() {
        return lastPageAccessed;
    }

    public void setLastPageAccessed() {
        this.lastPageAccessed = lastPageAccessed;
    }

    public boolean isInFirstWindow() {
        if(getPageSize() == 0) {
            return true;
        }
        return lastPageAccessed == 0;
    }

    public boolean isInLastWindow() {
        if(getPageSize() == 0) {
            return true;
        }
        return lastPageAccessed >= getTotalPageCount()-1;
    }

    public List<Integer> fetchPagesWindow() {
        List<Integer> ret = new ArrayList<Integer>(getPageSize());

        if(getTotalPageCount() == 1) {
            return ret;
        }
        for (int cnt = lastPageAccessed; cnt < getTotalPageCount(); cnt++) {
            ret.add(cnt);
        }
        return ret;
    }

}

File: web/src/main/scala/org/extrasapache/myfaces/codi/examples/ebean/support/data/PaginationController.scala
package org.extrasapache.myfaces.codi.examples.ebean.support.data

import java.util.concurrent.Future
import com.avaje.ebean.{Page, PagingList}

import scala.math._
import java.util._

/**
 *
 * @author Werner Punz (latest modification by $Author$)
 * @version $Revision$ $Date$
 *
 * This is our pagination controller, not most of the times
 * we can get away by using the Int -> int(java) autoboxing
 * but since scala does not know any Integer -> int autoboxing
 * we cannot get away the other way
 *
 * Also a null value enforces integer to be used
 */

class PaginationController[T](var _delegate: PagingList[T]) {
  //var _delegate: PagingList[T] = _
  var lastPageAccessed: Int = 0
  var pagingWindowSize: Int = 8

  def refresh() = _delegate.refresh

  def setFetchAhead(b: Boolean) = _delegate.setFetchAhead(b)

  def getFutureRowCount: Future[java.lang.Integer] = _delegate.getFutureRowCount

  def getAsList: java.util.List[T] = _delegate.getAsList

  def getPageSize: Int = _delegate.getPageSize

  def getTotalRowCount: Int = _delegate.getTotalRowCount

  def getTotalPageCount: Int = _delegate.getTotalPageCount

  def getCurrentPageRow: Int = _delegate.getPageSize * lastPageAccessed

  def getPage(i: Int): Page[T] = {
    lastPageAccessed = i
    _delegate.getPage(i)
  }

  def fetchPage: Page[T] = {
    _delegate.getPage(lastPageAccessed)
  }

  def isFirstPage: Boolean = lastPageAccessed == 0

  def isLastPage: Boolean = lastPageAccessed >= getTotalPageCount

  def isInFirstWindow: Boolean = (getPageSize == 0) || (lastPageAccessed < pagingWindowSize)

  def hasPreviousPagesWindow: Boolean = false

  def getFirstPageIdx: Int = 0

  def getLastPageIdx: Int = _delegate.getTotalPageCount - 1

  def getPreviousPage: Int = {
    val ret = max(0, lastPageAccessed - 1)
    ret
  }

  def getNextWindowIdx: Int = {
    if (isInLastWindow)
      (floor(lastPageAccessed.doubleValue / pagingWindowSize.doubleValue).intValue) * pagingWindowSize
    else
      (floor(lastPageAccessed.doubleValue / pagingWindowSize.doubleValue).intValue + 1) * pagingWindowSize
  }

  def getPreviousWindowIdx: Int = {
    if (isInFirstWindow)
      0
    else {
      (floor(lastPageAccessed.doubleValue / pagingWindowSize.doubleValue).intValue - 1) * pagingWindowSize
    }
  }

  def isInLastWindow: Boolean = {
    (getPageSize == 0) || (lastPageAccessed >= getTotalPageCount - pagingWindowSize)
  }

  def fetchPagesWindow: List[Int] = {
    val ret = new ArrayList[Int](getPageSize)
    if (getTotalPageCount != 1) {
      val pagingWindowPos = floor(lastPageAccessed.doubleValue / pagingWindowSize.doubleValue).intValue
      (pagingWindowPos until getTotalPageCount).foreach {
        ret.add(_)
      }
    }
    ret
  }

  def getPageAsList: List[T] = {
    fetchPage.getList
  }

  def getNextPage: Int = {
    max(0, min(lastPageAccessed, getTotalPageCount - 1))
  }

}